File: benchmark/python/binary_trees.py
import time
import sys

sys.setrecursionlimit(20000)

def bottom_up_tree(depth):
    if depth > 0:
        return (bottom_up_tree(depth - 1), bottom_up_tree(depth - 1))
    else:
        return (None, None)

def item_check(tree):
    if tree[0] is not None:
        return 1 + item_check(tree[0]) + item_check(tree[1])
    else:
        return 1

start = time.time() * 1000
max_depth = 12
min_depth = 4
stretch_depth = max_depth + 1

stretch_tree = bottom_up_tree(stretch_depth)
check = item_check(stretch_tree)

long_lived_tree = bottom_up_tree(max_depth)

for depth in range(min_depth, max_depth + 1, 2):
    iterations = 1 << (max_depth - depth + min_depth)
    check_sum = 0
    for i in range(iterations):
        check_sum += item_check(bottom_up_tree(depth))

long_lived_check = item_check(long_lived_tree)
end = time.time() * 1000

print(f"Result: {check}_{long_lived_check}")
print(f"Time: {end - start:.0f}ms")

File: benchmark/src/lua/binary_trees.lua
function bottom_up_tree(depth)
    if depth > 0 then
        return {bottom_up_tree(depth - 1), bottom_up_tree(depth - 1)}
    else
        return {}
    end
end

function item_check(tree)
    if tree[1] then
        return 1 + item_check(tree[1]) + item_check(tree[2])
    else
        return 1
    end
end

local start = os.clock() * 1000
local max_depth = 12
local min_depth = 4
local stretch_depth = max_depth + 1

local stretch_tree = bottom_up_tree(stretch_depth)
local check = item_check(stretch_tree)

local long_lived_tree = bottom_up_tree(max_depth)

for depth = min_depth, max_depth, 2 do
    local iterations = math.floor(2 ^ (max_depth - depth + min_depth))
    local check_sum = 0
    for i = 1, iterations do
        check_sum = check_sum + item_check(bottom_up_tree(depth))
    end
end

local long_lived_check = item_check(long_lived_tree)
local end_time = os.clock() * 1000

print("Result: " .. check .. "_" .. long_lived_check)
print("Time: " .. math.floor(end_time - start) .. "ms")
